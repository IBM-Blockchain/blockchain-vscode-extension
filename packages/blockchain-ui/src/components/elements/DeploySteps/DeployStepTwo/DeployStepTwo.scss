.deploy-step-two-notification {
    .bx--inline-notification__text-wrapper {
        flex-direction: column;
        align-items: flex-start !important;
    }
}
