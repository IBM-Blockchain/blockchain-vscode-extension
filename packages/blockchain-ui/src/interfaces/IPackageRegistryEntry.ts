interface IPackageRegistryEntry {
    name: string;
    path: string;
    version?: string;
    sizeKB: number;
}

export default IPackageRegistryEntry;
