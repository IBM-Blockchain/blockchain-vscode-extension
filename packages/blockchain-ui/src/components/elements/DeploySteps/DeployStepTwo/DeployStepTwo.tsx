import React, { Component } from 'react';
import { TextInput, FileUploader, Accordion, AccordionItem, Link } from 'carbon-components-react';
import IPackageRegistryEntry from '../../../../interfaces/IPackageRegistryEntry';

interface IProps {
    selectedPackage: IPackageRegistryEntry;
    onDefinitionNameChange: (data: any) => void;
    onDefinitionVersionChange: (data: any) => void;
    currentDefinitionName: string;
    currentDefinitionVersion: string;
}

interface DeployStepTwoState {
    definitionNameValue: string;
    definitionVersionValue: string;
}

class DeployStepTwo extends Component<IProps, DeployStepTwoState> {

    constructor(props: Readonly<IProps>) {
        super(props);

        this.state = {
            definitionNameValue: this.props.currentDefinitionName ? this.props.currentDefinitionName : this.props.selectedPackage.name,
            definitionVersionValue: this.props.currentDefinitionVersion ? this.props.currentDefinitionVersion : this.props.selectedPackage.version
        };

        this.handleDefinitionNameChange = this.handleDefinitionNameChange.bind(this);
        this.handleDefinitionVersionChange = this.handleDefinitionVersionChange.bind(this);

    }

    handleDefinitionNameChange(data: any): void {
        this.props.onDefinitionNameChange(data.target.value);
    }
    handleDefinitionVersionChange(data: any): void {
        this.props.onDefinitionVersionChange(data.target.value);
    }

    render(): JSX.Element {
        return (
            <>
                <div className='bx--row margin-bottom-06'>
                    <div className='bx--col-lg-10'>
                        A smart contract definition describes how the selected package will be deployed in this channel.
                        The package's name and version (where available) are copied across to the definition as defaults.
                        You may optionally change them.
                    </div>
                </div>
                <div className='bx--row'>
                    <h5>Smart contract definition</h5>
                </div>
                <div className='bx--row margin-bottom-06'>
                    <div className='bx--col-lg-5 margin-right-06'>
                        <TextInput id='nameInput' labelText='Definition name' defaultValue={this.state.definitionNameValue} onChange={this.handleDefinitionNameChange}></TextInput>
                    </div>
                    <div className='bx--col-lg-5'>
                        <TextInput id='versionInput' labelText='Definition version' defaultValue={this.state.definitionVersionValue} onChange={this.handleDefinitionVersionChange}></TextInput>
                    </div>
                </div>

                <div className='bx--row margin-bottom-06'>
                    <div className='bx--col-lg-5 margin-right-06'>
                        <div className='bx--file__container'>
                            <FileUploader
                                accept={[
                                '.json'
                                ]}
                                buttonKind='tertiary'
                                buttonLabel='Add file'
                                labelDescription='Default (1 endorser, any org) will be used unless you add your own.'
                                labelTitle='Endorsement policy'
                                size='default'
                            />
                        </div>
                    </div>
                    <div className='bx--col-lg-5'>
                        <div className='bx--file__container'>
                            <FileUploader
                                accept={[
                                '.json'
                                ]}
                                buttonKind='tertiary'
                                buttonLabel='Add file'
                                labelDescription='Default (implicit private data collections only) will be used unless you add your own.'
                                labelTitle='Collections configuration'
                                size='default'
                            />
                        </div>
                    </div>

                </div>
                <div className='bx--row'>
                    <div className='bx--col-lg-10'>
                        <Accordion>
                            <AccordionItem title={'Learn about other smart contract definition parameters (advanced)'}>
                                <p className='margin-bottom-05'>These developer tools automatically handle some parameters for simplied development.
                                    For a full explainabtion of definition parameters and Fabric v2.X smart contract lifecycle, see the <Link href='#'>Fabric documentation</Link>.
                                </p>
                                <p className='margin-bottom-05'>
                                    To deploy with full control of all parameters, please use the <Link href='#'>IBM Blockchain Platform Console</Link> or <Link href='#'>Hyperledger Fabric CLIs</Link>.
                                </p>
                                <p className='margin-bottom-05'>
                                    Read about how our simplified deployment handles the other parameters in the <Link href='#'>IBM Blockchain Platform documentation</Link>.
                                </p>
                            </AccordionItem>
                        </Accordion>
                    </div>
                </div>
            </>
        );
    }
}

export default DeployStepTwo;
