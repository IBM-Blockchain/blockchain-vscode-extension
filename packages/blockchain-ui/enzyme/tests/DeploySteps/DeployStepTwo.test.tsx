import React from 'react';
import renderer from 'react-test-renderer';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';
import { ReactWrapper, mount } from 'enzyme';

import DeployStepTwo from '../../../src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo';
import IPackageRegistryEntry from '../../../src/interfaces/IPackageRegistryEntry';
import { TextInput, FileUploaderItem } from 'carbon-components-react';

chai.should();
chai.use(sinonChai);

// tslint:disable no-unused-expression

describe('DeployStepTwo component', () => {
    let mySandBox: sinon.SinonSandbox;

    let onDefinitionNameChangeStub: sinon.SinonStub;
    let onDefinitionVersionChangeStub: sinon.SinonStub;
    let onCollectionChangeStub: sinon.SinonStub;

    const packageOne: IPackageRegistryEntry = {name: 'mycontract', version: '0.0.1', path: '/package/one', sizeKB: 9000};

    beforeEach(async () => {
        mySandBox = sinon.createSandbox();

        onDefinitionNameChangeStub = mySandBox.stub();
        onDefinitionVersionChangeStub = mySandBox.stub();
        onCollectionChangeStub = mySandBox.stub();
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    describe('render', () => {
        it('should render the expected snapshot', () => {
            const component: any = renderer
                .create(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />)
                .toJSON();
            expect(component).toMatchSnapshot();
        });

        it('should show the current definition name if available', () => {

            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='definitionName' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);

            const definitionInput: ReactWrapper<any> = component.find(TextInput).at(0);
            const nameInputProps: any = definitionInput.props();
            nameInputProps.defaultValue.should.equal('definitionName');
        });

        it('should show the current definition version if available', () => {

            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='0.0.3' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);

            const definitionInput: ReactWrapper<any> = component.find(TextInput).at(1);
            const nameInputProps: any = definitionInput.props();
            nameInputProps.defaultValue.should.equal('0.0.3');
        });

        it('should show uploaded collection', () => {
            const file: File = new File([], 'someFile');
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='0.0.3' currentCollectionFile={file} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const uploadItem: ReactWrapper<any> = component.find(FileUploaderItem).at(0);
            const html: string = uploadItem.html();
            html.includes(`<p class="bx--file-filename">${file.name}</p>`).should.equal(true);
        });

    });

    describe('handleDefinitionNameChange', () => {
        it('should handle definition name change', () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;
            const textInputData: any = {
                target: {
                    value: 'newDefinitionName'
                }
            };

            const isNameInvalidStub: sinon.SinonStub = mySandBox.stub(instance, 'isNameInvalid').returns(false);

            instance.handleDefinitionNameChange(textInputData);

            isNameInvalidStub.should.have.been.calledOnceWithExactly('newDefinitionName');
            onDefinitionNameChangeStub.should.have.been.calledOnceWithExactly('newDefinitionName', false);
        });
    });

    describe('handleDefinitionVersionChange', () => {
        it('should handle definition version change', () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;
            const textInputData: any = {
                target: {
                    value: '0.0.2'
                }
            };

            const isVersionInvalidStub: sinon.SinonStub = mySandBox.stub(instance, 'isVersionInvalid').returns(false);

            instance.handleDefinitionVersionChange(textInputData);

            isVersionInvalidStub.should.have.been.calledOnceWithExactly('0.0.2');
            onDefinitionVersionChangeStub.should.have.been.calledOnceWithExactly('0.0.2', false);
        });
    });

    describe('isNameInvalid', () => {
        it('should handle empty name', async () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;

            const result: boolean = instance.isNameInvalid('');
            result.should.equal(true);
        });

        it('should handle invalid name', async () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;

            const result: boolean = instance.isNameInvalid('invalid#name');
            result.should.equal(true);
        });

        it('should handle valid name', async () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;

            const result: boolean = instance.isNameInvalid('valid_contract-name-123');
            result.should.equal(false);
        });
    });

    describe('isVersionInvalid', () => {
        it('should handle empty version', async () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;

            const result: boolean = instance.isVersionInvalid('');
            result.should.equal(true);
        });

        it('should handle valid version', async () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;

            const result: boolean = instance.isVersionInvalid('0.0.1');
            result.should.equal(false);
        });
    });

    describe('collectionChange', () => {
        it('should pass a file to parent if selected', () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;

            const file: File = new File([], 'someFile');
            const event: any = {
                target: {
                    files: [
                        file
                    ]
                }
            };

            instance.collectionChange(event);

            onCollectionChangeStub.should.have.been.calledOnceWithExactly(file);
        });

        it('should not pass a file to parent if not selected', () => {
            const component: ReactWrapper<DeployStepTwo> = mount(<DeployStepTwo selectedPackage={packageOne} currentDefinitionName='' currentDefinitionVersion='' currentCollectionFile={undefined} onDefinitionNameChange={onDefinitionNameChangeStub} onDefinitionVersionChange={onDefinitionVersionChangeStub} onCollectionChange={onCollectionChangeStub} />);
            const instance: DeployStepTwo = component.instance() as DeployStepTwo;

            const event: any = {
                target: {
                    files: []
                }
            };

            instance.collectionChange(event);

            onCollectionChangeStub.should.not.have.been.called;
        });
    });
});
