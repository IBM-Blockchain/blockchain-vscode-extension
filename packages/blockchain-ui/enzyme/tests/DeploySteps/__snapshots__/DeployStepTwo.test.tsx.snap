// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployStepTwo component render should render the expected snapshot 1`] = `
Array [
  <div
    className="bx--row margin-bottom-06"
  >
    <div
      className="bx--col-lg-10"
    >
      A smart contract definition describes how the selected package will be deployed in this channel. The package's name and version (where available) are copied across to the definition as defaults. You may optionally change them.
    </div>
  </div>,
  <div
    className="bx--row margin-bottom-06"
  >
    <div
      className="bx--col-lg-11"
    >
      <h5>
        Smart contract definition
      </h5>
    </div>
    <div
      className="bx--col-lg-11"
    >
      <div
        className="bx--row"
      >
        <div
          className="bx--col-lg-7 bx--col-md-3 bx--col-sm-4"
        >
          <div
            className="bx--form-item bx--text-input-wrapper"
          >
            <label
              className="bx--label"
              htmlFor="nameInput"
            >
              Definition name
            </label>
            <div
              className="bx--text-input__field-wrapper"
              data-invalid={null}
            >
              <input
                className="bx--text-input bx--text__input"
                defaultValue="mycontract"
                disabled={false}
                id="nameInput"
                onChange={[Function]}
                onClick={[Function]}
                type="text"
              />
            </div>
          </div>
        </div>
        <div
          className="bx--col-lg-2 bx--col-md-1 bx--col-sm-0"
        />
        <div
          className="bx--col-lg-7 bx--col-md-3 bx--col-sm-4"
        >
          <div
            className="bx--form-item bx--text-input-wrapper"
          >
            <label
              className="bx--label"
              htmlFor="versionInput"
            >
              Definition version
            </label>
            <div
              className="bx--text-input__field-wrapper"
              data-invalid={null}
            >
              <input
                className="bx--text-input bx--text__input"
                defaultValue="0.0.1"
                disabled={false}
                id="versionInput"
                onChange={[Function]}
                onClick={[Function]}
                type="text"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bx--row margin-bottom-06"
  >
    <div
      className="bx--col-lg-11"
    >
      <div
        className="bx--row"
      >
        <div
          className="bx--col-lg-7 bx--col-md-3 bx--col-sm-4"
        >
          <div
            className="bx--form-item bx--text-input-wrapper"
          >
            <label
              className="bx--label"
              htmlFor="endorsementInput"
            >
              Endorsement policy
            </label>
            <div
              className="bx--form__helper-text"
            >
              Default (Channel/Application/Endorsement) policy will be used unless you specify your own.
            </div>
            <div
              className="bx--text-input__field-wrapper"
              data-invalid={null}
            >
              <input
                className="bx--text-input bx--text__input"
                defaultValue=""
                disabled={false}
                id="endorsementInput"
                onChange={[Function]}
                onClick={[Function]}
                placeholder="e.g. OR(\\"Org1MSP.member\\",\\"Org2MSP.member\\")"
                title="e.g. OR(\\"Org1MSP.member\\",\\"Org2MSP.member\\")"
                type="text"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bx--row margin-bottom-06"
  >
    <div
      className="bx--col-lg-11"
    >
      <div
        className="bx--row"
      >
        <div
          className="bx--col-lg-7 bx--col-md-3 bx--col-sm-4"
        >
          <div
            className="bx--file__container"
          >
            <div
              className="bx--form-item"
              defaultValue={Array []}
              onChange={[Function]}
              onClick={[Function]}
              size="default"
            >
              <strong
                className="bx--file--label"
              >
                Collections configuration
              </strong>
              <p
                className="bx--label-description"
              >
                Default (implicit private data collections only) will be used unless you add your own.
              </p>
              <label
                aria-disabled={false}
                className="bx--btn bx--btn--tertiary"
                htmlFor="id2"
                onChange={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                tabIndex={0}
              >
                <span
                  role="button"
                >
                  Add file
                </span>
              </label>
              <input
                accept={
                  Array [
                    ".json",
                  ]
                }
                className="bx--visually-hidden"
                disabled={false}
                id="id2"
                multiple={false}
                onChange={[Function]}
                onClick={[Function]}
                tabIndex="-1"
                type="file"
              />
              <div
                className="bx--file-container"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="bx--row"
  >
    <div
      className="bx--col-lg-10"
    >
      <ul
        className="bx--accordion bx--accordion--end"
      >
        <li
          className="bx--accordion__item"
          onAnimationEnd={[Function]}
        >
          <button
            aria-expanded={false}
            className="bx--accordion__heading"
            onClick={[Function]}
            onKeyDown={[Function]}
            title="Expand/Collapse"
            type="button"
          >
            <svg
              aria-label="Expand/Collapse"
              className="bx--accordion__arrow"
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="11,8 6,13 5.3,12.3 9.6,8 5.3,3.7 6,3"
              />
            </svg>
            <div
              className="bx--accordion__title"
            >
              Learn about other smart contract definition parameters (advanced)
            </div>
          </button>
          <div
            className="bx--accordion__content"
          >
            <p>
              These developer tools automatically handle other parameters for simplified deployment. Learn what values are used in our 
              <a
                className="bx--link"
                href="https://github.com/IBM-Blockchain/blockchain-vscode-extension"
              >
                documentation
              </a>
              .
            </p>
            <p
              className="margin-bottom-05"
            >
              Find out more information about the endorsement policy syntax 
              <a
                className="bx--link"
                href="https://hyperledger-fabric.readthedocs.io/en/master/endorsement-policies.html#endorsement-policy-syntax"
              >
                here
              </a>
              .
            </p>
            <p
              className="margin-bottom-05"
            >
              To deploy with full control of all parameters, please use the 
              <a
                className="bx--link"
                href="https://cloud.ibm.com/catalog/services/blockchain-platform"
              >
                IBM Blockchain Platform Console
              </a>
               or 
              <a
                className="bx--link"
                href="https://hyperledger-fabric.readthedocs.io/en/release-2.0/chaincode_lifecycle.html"
              >
                Hyperledger Fabric CLIs
              </a>
              .
            </p>
          </div>
        </li>
      </ul>
    </div>
  </div>,
]
`;
